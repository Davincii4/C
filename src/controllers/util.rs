use super::prelude::*;
use crate::util::errors::{forbidden, internal, AppError, AppResult};
use conduit_router::RequestParams;

/// The Origin header (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin)
/// is sent with CORS requests and POST requests, and indicates where the request comes from.
/// We don't want to accept authenticated requests that originated from other sites, so this
/// function returns an error if the Origin header doesn't match what we expect "this site" to
/// be: https://crates.io in production, or http://localhost:port/ in development.
pub fn verify_origin(req: &dyn RequestExt) -> AppResult<()> {
    let headers = req.headers();
    let allowed_origins = &req.app().config.allowed_origins;

    let bad_origin = headers
        .get_all(header::ORIGIN)
        .iter()
        .find(|value| !allowed_origins.contains(value));

    if let Some(bad_origin) = bad_origin {
        let error_message =
            format!("only same-origin requests can be authenticated. got {bad_origin:?}");
        return Err(internal(&error_message).chain(forbidden()));
    }
    Ok(())
}

pub trait RequestParamExt<'a> {
    fn param(self, key: &str) -> Option<&'a str>;
}

impl<'a> RequestParamExt<'a> for &'a (dyn RequestExt + 'a) {
    fn param(self, key: &str) -> Option<&'a str> {
        self.params().find(key)
    }
}
