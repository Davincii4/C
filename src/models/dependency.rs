use diesel::deserialize::{self, FromSql};
use diesel::pg::Pg;
use diesel::sql_types::{Integer, Text};

use crate::models::{Crate, Version};
use crate::schema::*;
use crates_io_index::DependencyKind as IndexDependencyKind;

#[derive(Identifiable, Associations, Debug, Queryable, QueryableByName)]
#[diesel(
    table_name = dependencies,
    check_for_backend(diesel::pg::Pg),
    belongs_to(Version),
    belongs_to(Crate),
)]
pub struct Dependency {
    pub id: i32,
    pub version_id: i32,
    pub crate_id: i32,
    pub req: String,
    pub optional: bool,
    pub default_features: bool,
    pub features: Vec<String>,
    pub target: Option<String>,
    pub kind: DependencyKind,
    pub explicit_name: Option<String>,
}

#[derive(Debug, QueryableByName)]
pub struct ReverseDependency {
    #[diesel(embed)]
    pub dependency: Dependency,
    #[diesel(sql_type = Integer)]
    pub crate_downloads: i32,
    #[diesel(sql_type = Text, column_name = crate_name)]
    pub name: String,
}

#[derive(Copy, Clone, Serialize, Deserialize, Debug, FromSqlRow)]
#[serde(rename_all = "lowercase")]
#[repr(u32)]
pub enum DependencyKind {
    Normal = 0,
    Build = 1,
    Dev = 2,
    // if you add a kind here, be sure to update `from_row` below.
}

impl From<IndexDependencyKind> for DependencyKind {
    fn from(dk: IndexDependencyKind) -> Self {
        match dk {
            IndexDependencyKind::Normal => DependencyKind::Normal,
            IndexDependencyKind::Build => DependencyKind::Build,
            IndexDependencyKind::Dev => DependencyKind::Dev,
        }
    }
}

impl From<DependencyKind> for IndexDependencyKind {
    fn from(dk: DependencyKind) -> Self {
        match dk {
            DependencyKind::Normal => IndexDependencyKind::Normal,
            DependencyKind::Build => IndexDependencyKind::Build,
            DependencyKind::Dev => IndexDependencyKind::Dev,
        }
    }
}

impl FromSql<Integer, Pg> for DependencyKind {
    fn from_sql(bytes: diesel::pg::PgValue<'_>) -> deserialize::Result<Self> {
        match <i32 as FromSql<Integer, Pg>>::from_sql(bytes)? {
            0 => Ok(DependencyKind::Normal),
            1 => Ok(DependencyKind::Build),
            2 => Ok(DependencyKind::Dev),
            n => Err(format!("unknown kind: {n}").into()),
        }
    }
}
