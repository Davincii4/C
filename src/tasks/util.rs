pub(super) mod advisory_lock;
