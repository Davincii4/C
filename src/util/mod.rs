pub use self::errors::{bad_request, human, internal, internal_error, CargoError, CargoResult};
pub use self::errors::{std_error, ChainError};
pub use self::io_util::{read_fill, LimitErrorReader, read_le_u32};
pub use self::json::{json_error, json_response, json_error_200};
pub use self::request_proxy::RequestProxy;

pub mod errors;
pub mod rfc3339;
mod io_util;
mod json;
mod request_proxy;
