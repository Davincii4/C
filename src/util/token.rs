use diesel::{deserialize::FromSql, pg::Pg, serialize::ToSql, sql_types::Bytea};
use rand::{distributions::Uniform, rngs::OsRng, Rng};
use sha2::{Digest, Sha256};

const TOKEN_LENGTH: usize = 32;

/// NEVER CHANGE THE PREFIX OF EXISTING TOKENS!!! Doing so will implicitly
/// revoke all the tokens, disrupting production users.
const TOKEN_PREFIX: &str = "cio";

#[derive(FromSqlRow, AsExpression, Clone, PartialEq, Eq)]
#[diesel(sql_type = Bytea)]
pub struct SecureToken {
    sha256: Vec<u8>,
}

impl SecureToken {
    pub(crate) fn generate() -> NewSecureToken {
        let plaintext = format!(
            "{}{}",
            TOKEN_PREFIX,
            generate_secure_alphanumeric_string(TOKEN_LENGTH)
        );
        let sha256 = Self::hash(&plaintext);

        NewSecureToken {
            plaintext,
            inner: Self { sha256 },
        }
    }

    pub(crate) fn parse(plaintext: &str) -> Option<Self> {
        // This will both reject tokens without a prefix and tokens of the wrong kind.
        if !plaintext.starts_with(TOKEN_PREFIX) {
            return None;
        }

        let sha256 = Self::hash(plaintext);
        Some(Self { sha256 })
    }

    pub fn hash(plaintext: &str) -> Vec<u8> {
        Sha256::digest(plaintext.as_bytes()).as_slice().to_vec()
    }
}

impl std::fmt::Debug for SecureToken {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str("SecureToken")
    }
}

impl ToSql<Bytea, Pg> for SecureToken {
    fn to_sql(&self, out: &mut diesel::serialize::Output<'_, '_, Pg>) -> diesel::serialize::Result {
        ToSql::<Bytea, Pg>::to_sql(&self.sha256, &mut out.reborrow())
    }
}

impl FromSql<Bytea, Pg> for SecureToken {
    fn from_sql(bytes: diesel::pg::PgValue<'_>) -> diesel::deserialize::Result<Self> {
        Ok(Self {
            sha256: FromSql::<Bytea, Pg>::from_sql(bytes)?,
        })
    }
}

pub(crate) struct NewSecureToken {
    plaintext: String,
    inner: SecureToken,
}

impl NewSecureToken {
    pub(crate) fn plaintext(&self) -> &str {
        &self.plaintext
    }

    #[cfg(test)]
    pub(crate) fn into_inner(self) -> SecureToken {
        self.inner
    }
}

impl std::ops::Deref for NewSecureToken {
    type Target = SecureToken;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

fn generate_secure_alphanumeric_string(len: usize) -> String {
    const CHARS: &[u8] = b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

    OsRng
        .sample_iter(Uniform::from(0..CHARS.len()))
        .map(|idx| CHARS[idx] as char)
        .take(len)
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generated_and_parse() {
        let token = SecureToken::generate();
        assert!(token.plaintext().starts_with(TOKEN_PREFIX));
        assert_eq!(
            token.sha256,
            Sha256::digest(token.plaintext().as_bytes()).as_slice()
        );

        let parsed = SecureToken::parse(token.plaintext()).expect("failed to parse back the token");
        assert_eq!(parsed.sha256, token.sha256);
    }

    #[test]
    fn test_parse_no_kind() {
        assert!(SecureToken::parse("nokind").is_none());
    }
}
