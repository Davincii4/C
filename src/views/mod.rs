// TODO: Move all encodable types here
// For now, just reexport

use std::collections::HashMap;
use std::fmt;
use std::str::FromStr;

use chrono::{NaiveDate, NaiveDateTime};
use semver;
use serde::{self, Deserialize, Deserializer, Serialize, Serializer};

pub use badge::EncodableBadge;

use util::errors::{human, CargoError, CargoResult};

#[derive(Serialize, Deserialize, Debug)]
pub struct EncodableCategory {
    pub id: String,
    pub category: String,
    pub slug: String,
    pub description: String,
    #[serde(with = "::util::rfc3339")] pub created_at: NaiveDateTime,
    pub crates_cnt: i32,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct EncodableCategoryWithSubcategories {
    pub id: String,
    pub category: String,
    pub slug: String,
    pub description: String,
    #[serde(with = "::util::rfc3339")] pub created_at: NaiveDateTime,
    pub crates_cnt: i32,
    pub subcategories: Vec<EncodableCategory>,
}

pub use crate_owner_invitation::{EncodableCrateOwnerInvitation, InvitationResponse};
pub use dependency::EncodableDependency;
pub use download::EncodableVersionDownload;

#[derive(Serialize, Deserialize, Debug)]
pub struct EncodableKeyword {
    pub id: String,
    pub keyword: String,
    #[serde(with = "::util::rfc3339")] pub created_at: NaiveDateTime,
    pub crates_cnt: i32,
}

pub use krate::EncodableCrate;
pub use owner::{EncodableOwner, EncodableTeam};
pub use token::EncodableApiTokenWithToken;
pub use user::{EncodablePrivateUser, EncodablePublicUser};
pub use version::{EncodableVersion, EncodableVersionLinks};

// TODO: Prefix many of these with `Encodable` then clean up the reexports
pub mod krate_publish;
pub use self::krate_publish::CrateDependency as EncodableCrateDependency;
pub use self::krate_publish::NewCrate as EncodableCrateUpload;

/// Information about whether this version built on the specified Rust version and target, as
/// uploaded by the `cargo publish-build-info` command.
#[derive(Debug, Serialize, Deserialize)]
pub struct EncodableVersionBuildInfoUpload {
    pub rust_version: EncodableRustChannelVersion,
    pub target: String,
    pub passed: bool,
}

/// Aggregated build info for a crate version grouped by Rust channel for front-end display
/// convenience.
#[derive(Debug, Serialize, Deserialize, Default)]
pub struct EncodableVersionBuildInfo {
    pub id: i32,
    pub ordering: HashMap<String, Vec<String>>,
    pub stable: HashMap<String, HashMap<String, bool>>,
    pub beta: HashMap<NaiveDate, HashMap<String, bool>>,
    pub nightly: HashMap<NaiveDate, HashMap<String, bool>>,
}

/// Describes a Rust version by its channel and the released version on that channel.
/// For use in describing what versions of Rust a particular crate version builds with.
/// Contains the original version string for inserting into the database.
#[derive(Debug)]
pub struct EncodableRustChannelVersion {
    raw: String,
    pub parsed: ParsedRustChannelVersion,
}

/// A pretty, minimal representation of a Rust version's channel and released version on that
/// channel. Namely, does not include the exact release hash.
#[derive(Debug)]
pub enum ParsedRustChannelVersion {
    Stable(semver::Version),
    Beta(NaiveDate),
    Nightly(NaiveDate),
}

impl fmt::Display for EncodableRustChannelVersion {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.raw)
    }
}

impl Serialize for EncodableRustChannelVersion {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_str(&self.raw)
    }
}

impl<'de> Deserialize<'de> for EncodableRustChannelVersion {
    fn deserialize<D: Deserializer<'de>>(d: D) -> Result<EncodableRustChannelVersion, D::Error> {
        let s = String::deserialize(d)?;

        Ok(EncodableRustChannelVersion {
            raw: s.clone(),
            parsed: ParsedRustChannelVersion::from_str(&s).map_err(serde::de::Error::custom)?,
        })
    }
}

impl FromStr for ParsedRustChannelVersion {
    type Err = Box<CargoError>;

    fn from_str(s: &str) -> CargoResult<Self> {
        // Recognized formats:
        // rustc 1.14.0 (e8a012324 2016-12-16)
        // rustc 1.15.0-beta.5 (10893a9a3 2017-01-19)
        // rustc 1.16.0-nightly (df8debf6d 2017-01-25)

        let pieces: Vec<_> = s.split(&[' ', '(', ')'][..])
            .filter(|s| !s.trim().is_empty())
            .collect();

        if pieces.len() != 4 {
            return Err(human(&format_args!(
                "rust_version `{}` not recognized; \
                 expected format like `rustc X.Y.Z (SHA YYYY-MM-DD)`",
                s
            )));
        }

        if pieces[1].contains("nightly") {
            Ok(ParsedRustChannelVersion::Nightly(
                NaiveDate::parse_from_str(pieces[3], "%Y-%m-%d")?,
            ))
        } else if pieces[1].contains("beta") {
            Ok(ParsedRustChannelVersion::Beta(NaiveDate::parse_from_str(
                pieces[3],
                "%Y-%m-%d",
            )?))
        } else {
            let v = semver::Version::parse(pieces[1])?;
            if v.pre.is_empty() {
                Ok(ParsedRustChannelVersion::Stable(v))
            } else {
                Err(human(&format_args!(
                    "rust_version `{}` not recognized as nightly, beta, or stable",
                    s
                )))
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;
    use super::*;
    use chrono::NaiveDate;
    use serde_json;

    #[test]
    fn category_dates_serializes_to_rfc3339() {
        let cat = EncodableCategory {
            id: "".to_string(),
            category: "".to_string(),
            slug: "".to_string(),
            description: "".to_string(),
            crates_cnt: 1,
            created_at: NaiveDate::from_ymd(2017, 1, 6).and_hms(14, 23, 11),
        };
        let json = serde_json::to_string(&cat).unwrap();
        assert!(
            json.as_str()
                .find(r#""created_at":"2017-01-06T14:23:11+00:00""#)
                .is_some()
        );
    }

    #[test]
    fn category_with_sub_dates_serializes_to_rfc3339() {
        let cat = EncodableCategoryWithSubcategories {
            id: "".to_string(),
            category: "".to_string(),
            slug: "".to_string(),
            description: "".to_string(),
            crates_cnt: 1,
            created_at: NaiveDate::from_ymd(2017, 1, 6).and_hms(14, 23, 11),
            subcategories: vec![],
        };
        let json = serde_json::to_string(&cat).unwrap();
        assert!(
            json.as_str()
                .find(r#""created_at":"2017-01-06T14:23:11+00:00""#)
                .is_some()
        );
    }

    #[test]
    fn keyword_serializes_to_rfc3339() {
        let key = EncodableKeyword {
            id: "".to_string(),
            keyword: "".to_string(),
            created_at: NaiveDate::from_ymd(2017, 1, 6).and_hms(14, 23, 11),
            crates_cnt: 0,
        };
        let json = serde_json::to_string(&key).unwrap();
        assert!(
            json.as_str()
                .find(r#""created_at":"2017-01-06T14:23:11+00:00""#)
                .is_some()
        );
    }

    #[test]
    fn version_serializes_to_rfc3339() {
        let ver = EncodableVersion {
            id: 1,
            krate: "".to_string(),
            num: "".to_string(),
            dl_path: "".to_string(),
            readme_path: "".to_string(),
            updated_at: NaiveDate::from_ymd(2017, 1, 6).and_hms(14, 23, 11),
            created_at: NaiveDate::from_ymd(2017, 1, 6).and_hms(14, 23, 12),
            downloads: 0,
            features: HashMap::new(),
            yanked: false,
            license: None,
            links: EncodableVersionLinks {
                dependencies: "".to_string(),
                version_downloads: "".to_string(),
                authors: "".to_string(),
            },
        };
        let json = serde_json::to_string(&ver).unwrap();
        assert!(
            json.as_str()
                .find(r#""updated_at":"2017-01-06T14:23:11+00:00""#)
                .is_some()
        );
        assert!(
            json.as_str()
                .find(r#""created_at":"2017-01-06T14:23:12+00:00""#)
                .is_some()
        );
    }
}
