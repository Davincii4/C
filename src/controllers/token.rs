use super::frontend_prelude::*;

use crate::models::ApiToken;
use crate::schema::api_tokens;
use crate::views::EncodableApiTokenWithToken;

use crate::auth::AuthCheck;
use crate::models::token::{CrateScope, EndpointScope};
use axum::response::IntoResponse;
use serde_json as json;

/// Handles the `GET /me/tokens` route.
pub async fn list(app: AppState, req: Parts) -> AppResult<Json<Value>> {
    conduit_compat(move || {
        let conn = app.db_read_prefer_primary()?;
        let auth = AuthCheck::only_cookie().check(&req, &conn)?;
        let user = auth.user();

        let tokens: Vec<ApiToken> = ApiToken::belonging_to(user)
            .filter(api_tokens::revoked.eq(false))
            .order(api_tokens::created_at.desc())
            .load(&*conn)?;

        Ok(Json(json!({ "api_tokens": tokens })))
    })
    .await
}

/// Handles the `PUT /me/tokens` route.
pub async fn new(app: AppState, req: BytesRequest) -> AppResult<Json<Value>> {
    conduit_compat(move || {
        /// The incoming serialization format for the `ApiToken` model.
        #[derive(Deserialize, Serialize)]
        struct NewApiToken {
            name: String,
            crate_scopes: Option<Vec<String>>,
            endpoint_scopes: Option<Vec<String>>,
        }

        /// The incoming serialization format for the `ApiToken` model.
        #[derive(Deserialize, Serialize)]
        struct NewApiTokenRequest {
            api_token: NewApiToken,
        }

        let new: NewApiTokenRequest = json::from_slice(req.body())
            .map_err(|e| bad_request(&format!("invalid new token request: {e:?}")))?;

        let name = &new.api_token.name;
        if name.is_empty() {
            return Err(bad_request("name must have a value"));
        }

        let conn = app.db_write()?;

        let auth = AuthCheck::default().check(&req, &conn)?;
        if auth.api_token_id().is_some() {
            return Err(bad_request(
                "cannot use an API token to create a new API token",
            ));
        }

        let user = auth.user();

        let max_token_per_user = 500;
        let count: i64 = ApiToken::belonging_to(user).count().get_result(&*conn)?;
        if count >= max_token_per_user {
            return Err(bad_request(&format!(
                "maximum tokens per user is: {max_token_per_user}"
            )));
        }

        let crate_scopes = new
            .api_token
            .crate_scopes
            .map(|scopes| {
                scopes
                    .into_iter()
                    .map(CrateScope::try_from)
                    .collect::<Result<Vec<_>, _>>()
            })
            .transpose()
            .map_err(|_err| bad_request("invalid crate scope"))?;

        let endpoint_scopes = new
            .api_token
            .endpoint_scopes
            .map(|scopes| {
                scopes
                    .into_iter()
                    .map(|scope| EndpointScope::try_from(scope.as_bytes()))
                    .collect::<Result<Vec<_>, _>>()
            })
            .transpose()
            .map_err(|_err| bad_request("invalid endpoint scope"))?;

        let api_token =
            ApiToken::insert_with_scopes(&conn, user.id, name, crate_scopes, endpoint_scopes)?;
        let api_token = EncodableApiTokenWithToken::from(api_token);

        Ok(Json(json!({ "api_token": api_token })))
    })
    .await
}

/// Handles the `DELETE /me/tokens/:id` route.
pub async fn revoke(app: AppState, Path(id): Path<i32>, req: Parts) -> AppResult<Json<Value>> {
    conduit_compat(move || {
        let conn = app.db_write()?;
        let auth = AuthCheck::default().check(&req, &conn)?;
        let user = auth.user();
        diesel::update(ApiToken::belonging_to(user).find(id))
            .set(api_tokens::revoked.eq(true))
            .execute(&*conn)?;

        Ok(Json(json!({})))
    })
    .await
}

/// Handles the `DELETE /tokens/current` route.
pub async fn revoke_current(app: AppState, req: Parts) -> AppResult<Response> {
    conduit_compat(move || {
        let conn = app.db_write()?;
        let auth = AuthCheck::default().check(&req, &conn)?;
        let api_token_id = auth
            .api_token_id()
            .ok_or_else(|| bad_request("token not provided"))?;

        diesel::update(api_tokens::table.filter(api_tokens::id.eq(api_token_id)))
            .set(api_tokens::revoked.eq(true))
            .execute(&*conn)?;

        Ok(StatusCode::NO_CONTENT.into_response())
    })
    .await
}
