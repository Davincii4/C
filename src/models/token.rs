mod scopes;

use chrono::NaiveDateTime;
use diesel::prelude::*;
use secrecy::SecretString;

pub use self::scopes::{CrateScope, EndpointScope};
use crate::models::User;
use crate::schema::api_tokens;
use crate::util::errors::{AppResult, InsecurelyGeneratedTokenRevoked};
use crate::util::rfc3339;
use crate::util::token::SecureToken;

/// The model representing a row in the `api_tokens` database table.
#[derive(Clone, Debug, PartialEq, Eq, Identifiable, Queryable, Associations, Serialize)]
#[diesel(belongs_to(User))]
pub struct ApiToken {
    pub id: i32,
    #[serde(skip)]
    pub user_id: i32,
    #[serde(skip)]
    token: SecureToken,
    pub name: String,
    #[serde(with = "rfc3339")]
    pub created_at: NaiveDateTime,
    #[serde(with = "rfc3339::option")]
    pub last_used_at: Option<NaiveDateTime>,
    #[serde(skip)]
    pub revoked: bool,
    /// `None` or a list of crate scope patterns (see RFC #2947)
    pub crate_scopes: Option<Vec<CrateScope>>,
    /// A list of endpoint scopes or `None` for the `legacy` endpoint scope (see RFC #2947)
    pub endpoint_scopes: Option<Vec<EndpointScope>>,
}

impl ApiToken {
    /// Generates a new named API token for a user
    pub fn insert(conn: &mut PgConnection, user_id: i32, name: &str) -> AppResult<CreatedApiToken> {
        Self::insert_with_scopes(conn, user_id, name, None, None)
    }

    pub fn insert_with_scopes(
        conn: &mut PgConnection,
        user_id: i32,
        name: &str,
        crate_scopes: Option<Vec<CrateScope>>,
        endpoint_scopes: Option<Vec<EndpointScope>>,
    ) -> AppResult<CreatedApiToken> {
        let token = SecureToken::generate();

        let model: ApiToken = diesel::insert_into(api_tokens::table)
            .values((
                api_tokens::user_id.eq(user_id),
                api_tokens::name.eq(name),
                api_tokens::token.eq(&*token),
                api_tokens::crate_scopes.eq(crate_scopes),
                api_tokens::endpoint_scopes.eq(endpoint_scopes),
            ))
            .get_result(conn)?;

        Ok(CreatedApiToken {
            plaintext: SecretString::from(token.plaintext().to_string()),
            model,
        })
    }

    pub fn find_by_api_token(conn: &mut PgConnection, token_: &str) -> AppResult<ApiToken> {
        use crate::schema::api_tokens::dsl::*;
        use diesel::{dsl::now, update};

        let token_ =
            SecureToken::parse(token_).ok_or_else(InsecurelyGeneratedTokenRevoked::boxed)?;

        let tokens = api_tokens
            .filter(revoked.eq(false))
            .filter(token.eq(&token_));

        // If the database is in read only mode, we can't update last_used_at.
        // Try updating in a new transaction, if that fails, fall back to reading
        conn.transaction(|conn| {
            update(tokens)
                .set(last_used_at.eq(now.nullable()))
                .get_result(conn)
        })
        .or_else(|_| tokens.first(conn))
        .map_err(Into::into)
    }
}

#[derive(Debug)]
pub struct CreatedApiToken {
    pub model: ApiToken,
    pub plaintext: SecretString,
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::NaiveDate;

    #[test]
    fn api_token_serializes_to_rfc3339() {
        let tok = ApiToken {
            id: 12345,
            user_id: 23456,
            token: SecureToken::generate().into_inner(),
            revoked: false,
            name: "".to_string(),
            created_at: NaiveDate::from_ymd_opt(2017, 1, 6)
                .unwrap()
                .and_hms_opt(14, 23, 11)
                .unwrap(),
            last_used_at: Some(
                NaiveDate::from_ymd_opt(2017, 1, 6)
                    .unwrap()
                    .and_hms_opt(14, 23, 12),
            )
            .unwrap(),
            crate_scopes: None,
            endpoint_scopes: None,
        };
        let json = serde_json::to_string(&tok).unwrap();
        assert_some!(json
            .as_str()
            .find(r#""created_at":"2017-01-06T14:23:11+00:00""#));
        assert_some!(json
            .as_str()
            .find(r#""last_used_at":"2017-01-06T14:23:12+00:00""#));
    }
}
