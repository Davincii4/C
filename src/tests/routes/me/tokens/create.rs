use crate::util::{RequestHelper, TestApp};
use cargo_registry::models::token::{CrateScope, EndpointScope};
use cargo_registry::models::ApiToken;
use cargo_registry::views::EncodableApiTokenWithToken;
use diesel::prelude::*;
use http::StatusCode;

static NEW_BAR: &[u8] = br#"{ "api_token": { "name": "bar" } }"#;

#[derive(Deserialize)]
struct NewResponse {
    api_token: EncodableApiTokenWithToken,
}

#[test]
fn create_token_logged_out() {
    let (_, anon) = TestApp::init().empty();
    anon.put("/api/v1/me/tokens", NEW_BAR).assert_forbidden();
}

#[test]
fn create_token_invalid_request() {
    let (_, _, user) = TestApp::init().with_user();
    let invalid = br#"{ "name": "" }"#;
    let response = user.put::<()>("/api/v1/me/tokens", invalid);
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_eq!(
        response.into_json(),
        json!({ "errors": [{ "detail": "invalid new token request: Error(\"missing field `api_token`\", line: 1, column: 14)" }] })
    );
}

#[test]
fn create_token_no_name() {
    let (_, _, user) = TestApp::init().with_user();
    let empty_name = br#"{ "api_token": { "name": "" } }"#;
    let response = user.put::<()>("/api/v1/me/tokens", empty_name);
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_eq!(
        response.into_json(),
        json!({ "errors": [{ "detail": "name must have a value" }] })
    );
}

#[test]
fn create_token_exceeded_tokens_per_user() {
    let (app, _, user) = TestApp::init().with_user();
    let id = user.as_model().id;
    app.db(|conn| {
        for i in 0..1000 {
            assert_ok!(ApiToken::insert(conn, id, &format!("token {i}")));
        }
    });
    let response = user.put::<()>("/api/v1/me/tokens", NEW_BAR);
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_eq!(
        response.into_json(),
        json!({ "errors": [{ "detail": "maximum tokens per user is: 500" }] })
    );
}

#[test]
fn create_token_success() {
    let (app, _, user) = TestApp::init().with_user();

    let json: NewResponse = user.put("/api/v1/me/tokens", NEW_BAR).good();
    assert_eq!(json.api_token.name, "bar");
    assert!(!json.api_token.token.is_empty());

    let tokens: Vec<ApiToken> =
        app.db(|conn| assert_ok!(ApiToken::belonging_to(user.as_model()).load(conn)));
    assert_eq!(tokens.len(), 1);
    assert_eq!(tokens[0].name, "bar");
    assert!(!tokens[0].revoked);
    assert_eq!(tokens[0].last_used_at, None);
    assert_eq!(tokens[0].crate_scopes, None);
    assert_eq!(tokens[0].endpoint_scopes, None);
}

#[test]
fn create_token_multiple_have_different_values() {
    let (_, _, user) = TestApp::init().with_user();
    let first: NewResponse = user.put("/api/v1/me/tokens", NEW_BAR).good();
    let second: NewResponse = user.put("/api/v1/me/tokens", NEW_BAR).good();

    assert_ne!(first.api_token.token, second.api_token.token);
}

#[test]
fn create_token_multiple_users_have_different_values() {
    let (app, _, user1) = TestApp::init().with_user();
    let first_token: NewResponse = user1.put("/api/v1/me/tokens", NEW_BAR).good();

    let user2 = app.db_new_user("bar");
    let second_token: NewResponse = user2.put("/api/v1/me/tokens", NEW_BAR).good();

    assert_ne!(first_token.api_token.token, second_token.api_token.token);
}

#[test]
fn cannot_create_token_with_token() {
    let (_, _, _, token) = TestApp::init().with_token();
    let response = token.put::<()>(
        "/api/v1/me/tokens",
        br#"{ "api_token": { "name": "baz" } }"#,
    );
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_eq!(
        response.into_json(),
        json!({ "errors": [{ "detail": "cannot use an API token to create a new API token" }] })
    );
}

#[test]
fn create_token_with_scopes() {
    let (app, _, user) = TestApp::init().with_user();

    let json = json!({
        "api_token": {
            "name": "bar",
            "crate_scopes": ["tokio", "tokio-*"],
            "endpoint_scopes": ["publish-update"],
        }
    });

    let json: NewResponse = user
        .put("/api/v1/me/tokens", &serde_json::to_vec(&json).unwrap())
        .good();
    assert_eq!(json.api_token.name, "bar");
    assert!(!json.api_token.token.is_empty());

    let tokens: Vec<ApiToken> =
        app.db(|conn| assert_ok!(ApiToken::belonging_to(user.as_model()).load(conn)));
    assert_eq!(tokens.len(), 1);
    assert_eq!(tokens[0].name, "bar");
    assert!(!tokens[0].revoked);
    assert_eq!(tokens[0].last_used_at, None);
    assert_eq!(
        tokens[0].crate_scopes,
        Some(vec![
            CrateScope::try_from("tokio").unwrap(),
            CrateScope::try_from("tokio-*").unwrap()
        ])
    );
    assert_eq!(
        tokens[0].endpoint_scopes,
        Some(vec![EndpointScope::PublishUpdate])
    );
}

#[test]
fn create_token_with_null_scopes() {
    let (app, _, user) = TestApp::init().with_user();

    let json = json!({
        "api_token": {
            "name": "bar",
            "crate_scopes": null,
            "endpoint_scopes": null,
        }
    });

    let json: NewResponse = user
        .put("/api/v1/me/tokens", &serde_json::to_vec(&json).unwrap())
        .good();
    assert_eq!(json.api_token.name, "bar");
    assert!(!json.api_token.token.is_empty());

    let tokens: Vec<ApiToken> =
        app.db(|conn| assert_ok!(ApiToken::belonging_to(user.as_model()).load(conn)));
    assert_eq!(tokens.len(), 1);
    assert_eq!(tokens[0].name, "bar");
    assert!(!tokens[0].revoked);
    assert_eq!(tokens[0].last_used_at, None);
    assert_eq!(tokens[0].crate_scopes, None);
    assert_eq!(tokens[0].endpoint_scopes, None);
}

#[test]
fn create_token_with_empty_crate_scope() {
    let (_, _, user) = TestApp::init().with_user();

    let json = json!({
        "api_token": {
            "name": "bar",
            "crate_scopes": ["", "tokio-*"],
            "endpoint_scopes": ["publish-update"],
        }
    });

    let response = user.put::<()>("/api/v1/me/tokens", &serde_json::to_vec(&json).unwrap());
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_eq!(
        response.into_json(),
        json!({ "errors": [{ "detail": "invalid crate scope" }] })
    );
}

#[test]
fn create_token_with_invalid_endpoint_scope() {
    let (_, _, user) = TestApp::init().with_user();

    let json = json!({
        "api_token": {
            "name": "bar",
            "crate_scopes": ["tokio", "tokio-*"],
            "endpoint_scopes": ["crash"],
        }
    });

    let response = user.put::<()>("/api/v1/me/tokens", &serde_json::to_vec(&json).unwrap());
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_eq!(
        response.into_json(),
        json!({ "errors": [{ "detail": "invalid endpoint scope" }] })
    );
}
