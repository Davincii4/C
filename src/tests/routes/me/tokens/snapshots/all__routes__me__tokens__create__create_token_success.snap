---
source: src/tests/routes/me/tokens/create.rs
expression: response.into_json()
---
api_token:
  crate_scopes: ~
  created_at: "[datetime]"
  endpoint_scopes: ~
  id: "[id]"
  last_used_at: "[datetime]"
  name: bar
  token: "[token]"

